engine/api/timeline.go

```diff
@@ -18,27 +19,102 @@ func (api *API) getTimelineHandler() Handler {
 			return sdk.WrapError(errS, "getTimelineHandler> Invalid format for current item")
 		}
 
+		// Get workflow to mute
 		timelineFilter, errT := user.LoadTimelineFilter(api.mustDB(), u)
 		if errT != nil {
 			return sdk.WrapError(errT, "getTimelineHandler> Unable to load timeline filter")
 		}
 
-		if timelineFilter.AllProjects {
-			timelineFilter.Projects = make([]sdk.ProjectFilter, 0, len(u.Permissions.ProjectsPerm))
-			for k := range u.Permissions.ProjectsPerm {
-				timelineFilter.Projects = append(timelineFilter.Projects, sdk.ProjectFilter{Key: k})
+		allPerm := make(map[string][]string)
+
+		jobs := make(chan timelineFilterJob, 50)
+		results := make(chan timelineFilterJob, 50)
+
+		for w := 1; w <= 3; w++ {
+			go haveToFilter(jobs, results, timelineFilter)
+		}
+
+		for k := range u.Permissions.WorkflowsPerm {
+			keySplitted := strings.Split(k, "/")
+			pKey := keySplitted[0]
+			wName := keySplitted[1]
+			j := timelineFilterJob{
+				Key:          pKey,
+				WorkflowName: wName,
+			}
+			jobs <- j
+		}
+		close(jobs)
+
+		for a := 0; a < len(u.Permissions.WorkflowsPerm); a++ {
+			item := <-results
+			if !item.Add {
+				continue
+			}
+			workflows, ok := allPerm[item.Key]
+			if !ok {
+				workflows = make([]string, 0, 1)
 			}
+			workflows = append(workflows, item.WorkflowName)
+			allPerm[item.Key] = workflows
 		}
 
-		filter := sdk.EventFilter{
+		request := sdk.EventFilter{
 			CurrentItem: currentItem,
-			Filter:      timelineFilter,
+			Filter: sdk.TimelineFilter{
+				Projects: make([]sdk.ProjectFilter, 0, len(allPerm)),
+			},
+		}
+		for k, v := range allPerm {
+			pFilter := sdk.ProjectFilter{
+				Key:           k,
+				WorkflowNames: v,
+			}
+			request.Filter.Projects = append(request.Filter.Projects, pFilter)
 		}
 
-		events, err := event.GetEvents(api.mustDB(), api.Cache, filter)
+		events, err := event.GetEvents(api.mustDB(), api.Cache, request)
 		if err != nil {
 			return sdk.WrapError(err, "getTimelineHandler> Unable to load events")
 		}
 		return WriteJSON(w, events, http.StatusOK)
 	}
 }
+
+type timelineFilterJob struct {
+	Key          string
+	WorkflowName string
+	Add          bool
+}
+
+func haveToFilter(jobs <-chan timelineFilterJob, results chan<- timelineFilterJob, filter sdk.TimelineFilter) {
+	for j := range jobs {
+		projectFound := false
+		workflowFound := false
+		insert := false
+	projLoop:
+		for _, pf := range filter.Projects {
+			if pf.Key == j.Key {
+				projectFound = true
+				for _, wName := range pf.WorkflowNames {
+					if wName == j.WorkflowName {
+						workflowFound = true
```

Comment:
ineffectual assignment to `workflowFound`

Reply:
he's right